assert/assert_test.go

```diff
@@ -0,0 +1,221 @@
+package assert_test
+
+import (
+	"fmt"
+	"math"
+	"testing"
+
+	"github.com/madlambda/spells/assert"
+)
+
+func TestPartial(t *testing.T) {
+	type testcase struct {
+		name   string
+		obj    interface{}
+		target interface{}
+		fail   bool
+	}
+
+	for _, tc := range []testcase{
+		{
+			name:   "same numbers",
+			obj:    1,
+			target: 1,
+		},
+		{
+			name:   "different int64 numbers - bigger than int32",
+			obj:    int64(math.MaxInt32 + 1),
+			target: int64(math.MaxInt32 + 2),
+			fail:   true,
+		},
+		{
+			name:   "same uint64 numbers - bigger than int64",
+			obj:    uint64(9223372036854775807),
+			target: uint64(9223372036854775807 + 1),
+			fail:   true,
+		},
+		{
+			name:   "numbers mismatch",
+			obj:    1,
+			target: 0,
+			fail:   true,
+		},
+		{
+			name:   "same bool",
+			obj:    true,
+			target: true,
+		},
+		{
+			name:   "bool mismatch",
+			obj:    true,
+			target: false,
+			fail:   true,
+		},
+		{
+			name:   "same floats",
+			obj:    1.2,
+			target: 1.2,
+		},
+		{
+			name:   "different floats",
+			obj:    1.3,
+			target: 1.2,
+			fail:   true,
+		},
+		{
+			name:   "same empty struct",
+			obj:    struct{}{},
+			target: struct{}{},
+		},
+		{
+			name: "different number of fields",
+			obj: struct {
+				A int
+			}{},
+			target: struct{}{},
+			fail:   true,
+		},
+		{
+			name: "different field types",
+			obj: struct {
+				A int
+			}{},
+			target: struct {
+				B string
+			}{},
+			fail: true,
+		},
+		{
+			name: "same struct types different field names",
+			obj: struct {
+				A int
+			}{},
+			target: struct {
+				B int
+			}{},
+			fail: true,
+		},
+		{
+			name: "same struct value",
+			obj: struct {
+				A int
+			}{1},
+			target: struct {
+				A int
+			}{1},
+		},
+		{
+			name: "different struct field value",
+			obj: struct {
+				A string
+			}{"test"},
+			target: struct {
+				A string
+			}{"test2"},
+			fail: true,
+		},
+		{
+			obj: struct {
+				A string
+			}{"test"},
+			target: struct {
+				A string
+			}{"test"},
+			name: "same struct field value",
+		},
+		{
+			name: "field contains",
```

Comment:
hmm I thought that we were going just with ignoring struct fields that have 0 value on the target, but we are also doing contains on the strings ? will it also work like that for slices ? (like testing subsets ?)

I think we are missing a test for when the target has a zero value, from what we discussed before I imagine that zero values on the target are ignored.

Reply:
yeah I added contains in the strings because this is super super useful to avoid hacks such as fixing the `wanted` strings by prepending the sandbox.RootDir of filenames in the error message.. and other hacks.. most of the time I want to check if the field contains part of the string, not it all.
I think the name makes it clear it's a Partial() match, even in the specific fields, no?

Ah yes, I did the same kind of implementation for for slices and maps, the `target` drives what's going to be asserted in the object.

Still missing a good docs.
WDYT?

Reply:
I think it is a good idea, was just surprised at first because I was thinking just on the zero value scenario, but like you said, in tests it is quite usual to do matching and the name of the function is already about Partial things.

Good stuff :rocket: